go/mysql/conn.go

```diff
@@ -180,9 +180,24 @@ type Conn struct {
 
 	// PrepareData is the map to use a prepared statement.
 	PrepareData map[uint32]*PrepareData
+
+	// CursorState serves as a buffer for the current result set from the last COM_STMT_EXECUTE
+	// Rows are read from here upon receiving COM_STMT_FETCH
+	// This is basically a server-side cursor implementation
+	cs *cursorState
+}
+
+type cursorState struct {
+	stmtID  uint32
+	pending *sqltypes.Result
+	noData bool
+
+	next  chan *sqltypes.Result
+	done  chan error
+	quit  chan error
```

Comment:
As written we don't currently cancel the query anywhere by sending on this channel. As mentioned below though, I think we actually need to sometimes. When we do do that, we probably need to select like:

```
select {
  conn.cs.quit <- errToCancelTheQueryHere:
  errWhichWonTheRace := <- conn.cs.done:
}
```

in order to avoid deadlocking...